Fix fragment bugs

leoyvens opened the pull request:
This fixes issues in how prefetch handles redundant selections in fragments. The changes are mostly around field collection, which will now have a single group of fields for response keys of objects, and for response keys of interfaces it will have a group of fields for the interface and one for each concrete type that appears in a type condition in the query.

Field collection in `execution.rs` was also refactored, even though it's not directly related to the issues.

This also changes where field collection happens, to make selecting only the relevant fields from the DB in a future patch easier.

Resolves #1816.

leoyvens commented on graphql/src/store/prefetch.rs:
```diff
@@ -506,38 +466,15 @@ fn execute_root_selection_set(
-        if sast::get_field(query_type, &name).is_some() {
-            data_set.items.extend(selections)
-        }
-    }
```
I removed this explicit separation of the introspection fields, because introspection fields will be ignored by field collection anyways.

leoyvens replied:
We actually check that in too many places. We have proper validation for `UnknownField` errors in `validate_fields`, so I've removed all other checks for these and replaced them with unwraps.

lutter commented on graphql/src/store/prefetch.rs:
```diff
@@ -658,17 +602,73 @@ fn execute_selection_set<'a>(
+    fn collect_field(&mut self, type_condition: ObjectOrInterface<'a>, field: &'a q::Field) {
+        match type_condition {
+            ObjectOrInterface::Interface(i) => {
+                self.interface_cond = Some(i);
```
Why is it ok to clobber a potentially already existing `Some(_)` here?

leoyvens replied:
I've added an assert that if this assigns twice, it will assign the same value. It's not a very obvious invariant but `collect_fields` will never call this for two different interface types.

lutter commented on graphql/src/store/prefetch.rs:
```diff
@@ -26,61 +28,23 @@ lazy_static! {
-        }
-    }
+#[derive(Clone, Debug)]
+struct ObjectCondition<'a>(&'a s::ObjectType);
```
Would be good to comment that this is just an `ObjectType` that implements `Hash` and `Eq`

leoyvens replied:
I've added a comment.

lutter commented on graphql/src/store/prefetch.rs:
```diff
@@ -567,66 +504,73 @@ fn execute_selection_set<'a>(
+                    None => true,
+                    Some(relevant_set) => relevant_set.contains(c.0.name.as_str()),
+                })
+                .map(|(c, f)| (ObjectOrInterface::Object(c.0), f)),
```
There's a lot going on here, even though it's just about getting an iterator over all `(TypeCondition, Vec<Field>)`  pairs in the `collected_fields`. Could constructing the iterator be a function on `CollectedResponseKey`?

leoyvens replied:
I moved this into an `IntoIterator` implementation for `CollectedResponseKey`, now that we don't need the `relevant_set` anymore.

lutter commented on graphql/src/store/prefetch.rs:
```diff
@@ -658,17 +602,73 @@ fn execute_selection_set<'a>(
+    // concrete types, effectively merging the selection sets.
+    // See also: e0d6da3e-60cf-41a5-b83c-b60a7a766d4a
+    for collected_response_key in grouped_fields.values_mut() {
+        for concrete_type_fields in collected_response_key.concrete_types.values_mut() {
```
This `for` loop should be a function `merge_interface_fields` or maybe just `finish` on `CollectedResponseKey`

leoyvens replied:
I don't think that would gain us much clarity, and that method would only ever be called here, so I'm fine with how this is now.

lutter commented on graphql/src/store/prefetch.rs:
```diff
@@ -699,101 +699,77 @@ fn collect_fields<'a>(
-                                        .or_default()
-                                        .append(&mut group);
-                                }
+                    let fragment_ty = {
```
The body of this match arm is the same (I think) as most of the body of the previous arm; could this be factored into a helper function to make it clearer that we do almost the same thing for the two kinds of fragments, and what the differences are in how we treat them?

leoyvens replied:
There is now a `collect_fragment` function.

leoyvens commented on graphql/src/execution/execution.rs:
```diff
@@ -556,58 +570,27 @@ pub fn collect_fields<'a>(
-                                        .append(&mut fragment_group);
-                                }
-                            });
+                    if visited_fragments.insert(&spread.fragment_name) {
```
Interesting observation. If we were to check for this, it would be checked in a validation step and not here. But I don't see any reason to bother with not supporting cycles when they can be supported just fine. I've added a note here.

leoyvens commented on graphql/src/store/prefetch.rs:
```diff
@@ -717,76 +679,59 @@ fn collect_fields<'a>(
+                            object_or_interface_by_name(schema, ty_name).ok_or_else(|| {
+                                QueryExecutionError::NamedTypeError(ty_name.clone())
+                            })?
+                        };
```
Among the changes made there is now a `collect_fragment` function.

leoyvens commented on graphql/src/execution/execution.rs:
```diff
@@ -540,11 +540,25 @@ pub fn collect_fields<'a>(
     object_type: &s::ObjectType,
     selection_sets: impl Iterator<Item = &'a q::SelectionSet>,
-    visited_fragments: Option<HashSet<&'a q::Name>>,
 ) -> IndexMap<&'a String, Vec<&'a q::Field>> {
```
I've added a comment here.

leoyvens commented on graphql/src/store/prefetch.rs:
```diff
@@ -717,76 +679,59 @@ fn collect_fields<'a>(
+                        };
+
+                        if fragment_ty.is_object() {
+                            type_condition = fragment_ty
```
`type_condition` was actually never read from after this assignment, I've revamped the logic around this.

leoyvens commented on graphql/src/store/prefetch.rs:
```diff
@@ -411,24 +372,23 @@ impl<'a> Join<'a> {
+            //
+            // This `insert` will overwrite in the case where the response key occurs both at the
+            // interface level and in concrete types. The interface is always joined first, and may
+            // then be overwritten by the merged selection set under. the concrete type condition.
```
I've reworded this a bit to try to make it clearer. But yes, we continue to separate queries for separate type conditions returned by `collect_fields`.

leoyvens commented on graphql/src/store/prefetch.rs:
```diff
@@ -567,66 +504,73 @@ fn execute_selection_set<'a>(
+    }
+
+    // Skip dead fragments as an optimization
+    let relevant_concrete_types: Option<HashSet<String>> = if parents[0].entity.is_empty() {
```
I've improved this function to fetch and join only for parents which match the type condition. This ended up being necessary for correctness and is also a more general optimization than this was, so this check was removed.

leoyvens commented on graph/src/data/graphql/object_or_interface.rs:
```diff
@@ -0,0 +1,87 @@
+use std::collections::BTreeMap;
+
+#[derive(Copy, Clone, Debug)]
+pub enum ObjectOrInterface<'a> {
```
This got moved into the `graph` crate to be used in validation.

leoyvens commented:
@lutter Thanks for the review. I added tests for nested interface fragments which were still buggy. This required changing multiple aspects of this PR, so it will need a re-review (recommend reading with whitespace diff disabled). Though many of your points were still valid and I addressed them.

roienatan commented:
What is the status of this?

dkent600 commented:
@leoyvens Any ETA on release of this?  Thanks

lutter commented on graphql/src/store/prefetch.rs:
```diff
@@ -506,295 +453,309 @@ fn execute_root_selection_set(
 }
 
+#[derive(Default)]
+struct CollectedResponseKey<'a> {
```
Just to make sure I understand what this is (and might be good as a comment): a `CollectedResponseKey` contains a map of the `ObjectType` to the fields we fetch for that object type. If we actually fetch fields for an interface type, i.e., `iface_cond` is `Some(_)`, `iface_fields` contains the fields we fetch from the interface. Do the fields in `obj_types` already contain these fields or are they solely the type-specific fields we fetch? I'll probably find the answer to that question when I read `collect_field`, but it would be nice if it was stated explicitly.

leoyvens replied:
I've added a comment.

lutter commented on graphql/src/store/prefetch.rs:
```diff
@@ -506,295 +453,309 @@ fn execute_root_selection_set(
+            // each type in the intersection between the root interface and the
+            // interface in the fragment type condition.
+            let types_for_interface = ctx.query.schema.types_for_interface();
+            let root_tys = &types_for_interface[outer_type_condition.name()];
```
Is it possible that `outer_type_condition` is an object type, but `fragment_ty` is an interface? I think that would lead to a panic here.

leoyvens replied:
The comment states that `This is an interface fragment in the root selection for an interface`. It's a bit subtle, but if the outer type is an object, we consider any condition to be redundant and therefore don't enter this if arm.

lutter reviewed (approved):
Looks great! Feel free to merge after addressing the two comments
